src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp

```diff
@@ -178,19 +175,19 @@ void ShenandoahControlThread::run_service() {
 
       // Report current free set state at the end of cycle, whether
       // it is a normal completion, or the abort.
-      {
-        ShenandoahHeapLocker locker(heap->lock());
-        heap->free_set()->log_status();
+      heap->free_set()->log_status_under_lock();
 
+      {
         // Notify Universe about new heap usage. This has implications for
         // global soft refs policy, and we better report it every time heap
         // usage goes down.
+        ShenandoahHeapLocker locker(heap->lock());
         heap->update_capacity_and_used_at_gc();
-
-        // Signal that we have completed a visit to all live objects.
-        heap->record_whole_heap_examined_timestamp();
       }
```

Comment:
Backport of changes look sgood, but highlights the lock-unlock-lock-unlock awkwardness. I suppose the mitigating factor is that in most cases the first one doesn't usually lock because free set log status is hopefully disabled in most production scenarios.

Reply:
Yes, lock/unlock is not ideal when freeset logging is enabled. However, this arrangement make it clear that `update_capacity_and_used_at_gc` needs the lock separately.